module-network.tf

```diff
@@ -0,0 +1,187 @@
+# Copyright (c) 2017, 2023 Oracle Corporation and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+data "oci_core_vcns" "all" {
+  count          = var.create_vcn ? 0 : 1
+  compartment_id = local.compartment_id
+  display_name   = var.vcn_name
+  state          = "AVAILABLE"
+}
+
+data "oci_core_vcn" "oke" {
+  count  = var.create_vcn ? 0 : 1
+  vcn_id = var.vcn_id
+}
+
+data "oci_core_nat_gateways" "all" {
+  compartment_id = local.compartment_id
+  display_name   = var.nat_gateway_display_name
+  state          = "AVAILABLE"
+  vcn_id         = local.vcn_id
+}
+
+data "oci_core_internet_gateways" "all" {
+  compartment_id = local.compartment_id
+  display_name   = var.internet_gateway_display_name
+  state          = "AVAILABLE"
+  vcn_id         = local.vcn_id
+}
+
+data "oci_core_route_tables" "all" {
+  compartment_id = local.compartment_id
+  state          = "AVAILABLE"
+  vcn_id         = local.vcn_id
+}
+
+locals {
+  vcn_id = (var.create_vcn ? one(module.vcn[*].vcn_id)
+    : coalesce(var.vcn_id, "none") != "none" ? var.vcn_id
+    : one(one(data.oci_core_vcns.all[*].virtual_networks)[*].id)
```

Comment:
this is assuming that a vcn with the same display name will be the one you're looking for. However, as mentioned above, `display_name` is not unique.

Reply:
That's true - it's making a best effort at that point to find something available rather than fail (e.g. in a lot of cases there may just be 1) - should we require instead?

Reply:
Yup, as mentioned above. Either the user accepts creating the vcn here or provide a valid vcn ocid. By valid, this means they have the necessary permission etc.

Reply:
Done.